docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,188 @@
+
+Currently, CAPZ handles `2.` with Azure API calls, which requires ensuring dependencies between Azure resources are met and maintaining state to asynchronously handle long-running operations. Because ASO's primary objective is to handle `2.`, CAPZ could then more declaratively specify all of the infrastructure needed for a given Cluster API resource and let ASO handle the finer details of ensuring that desired state is achieved in Azure. By then primarily focusing on `1.`, CAPZ reduces its overall code maintenance burden and affirms ASO as the canonical way to manage Azure infrastructure from Kubernetes.
+
+In addition, the [Azure Go SDK](https://github.com/Azure/azure-sdk-for-go) has recently deprecated its set of APIs that CAPZ currently uses in favor of a new set of backwards-incompatible APIs. Because the deprecated Azure Go SDK APIs are no longer receiving updates to adopt new Azure API versions, CAPZ must adopt the new APIs to be able to expose the newest features in Azure to its users. Because ASO already uses the new APIs, adopting ASO in CAPZ accomplishes the goal of migrating off of the deprecated APIs.
+
+### Goals
+
+- Logically separate transformation of Cluster API resources to infrastructure definitions from reconciliation of that infrastructure in Azure
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
+- Determine how ASO will be installed on Cluster API management clusters
+
+### Non-Goals/Future Work
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Reconcile long-running Azure operations asynchronously by waiting for
+  Kubernetes watch events from changes to the `status` of ASO resources instead
+  of periodically checking if the operation is complete
+- Reconcile different Azure resource types for the same workload cluster in parallel
+- Allow users to interact with ASO resources directly, out-of-band with CAPZ
+- Adopt Azure Workload Identity to authenticate with Azure APIs
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Users of new Azure features
+
+As a CAPZ user I want to take advantage of the latest and greatest features in Azure only available in the newest Azure API versions in my clusters managed by CAPZ. This would allow me to adopt CAPZ for a wider range of use cases and continue evolving my CAPZ clusters alongside the Azure API.
+
+#### Story 2 - Developer experience
+
+As a CAPZ developer I would like to only need to interact with the Kubernetes API server using Custom Resource Definitions representing Azure resources rather than having to deal with the impedance of interacting with the Azure APIs. This would enable me to focus more on the core value of CAPZ, transforming Cluster API resource definitions to Azure resource definitions.
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
```

Comment:
My recommendation is to not install using the multitenant helm chart. Check [this](https://azure.github.io/azure-service-operator/introduction/multitenant-deployment/#single-operator-multitenancy-default) out to understand ASOs multitenancy options, but basically there are 2:

1. The default installation method, which is that a single operator watches all namespaces and can do multitenancy using namespace-scoped or resource-scoped credentials.
2. The multitenant installation method, which really means "multiple instances of ASO, each watching a subset of namespaces"

The reason to pick 2 is if you're security conscious and you don't trust ASOs standard multitenancy (which results in the single operator pod being a possible attack risk to exfiltrate all of your secrets since it has permission to read them all). Or if you're concerned about a bug in ASOs multitenancy. There's a lot of complexity you adopt when you use approach 2 though because there's still only a single set of CRDs (cluster scope resource), so you need to coordinate operator upgrades across N installs of the ASO in your cluster.

My understanding is that CAPZs multitenancy model is basically ASOs "default model" - a single CAPZ operator watches all namespaces and reads the credentials associated with each given cluster. So that maps much better onto the first model than the second.

Reply:
Great feedback, thanks! I've updated this section and tried to capture this here.